### ISSUE TITLE
Add batched torch.combinations

### ISSUE THREAD

**Issue by n2cholas** [pytorch/pytorch#40375]
## 🚀 Feature
I would like `torch.combinations` to support a batch of vectors. So, the signature would change from: `torch.combinations(input, r=2, with_replacement=False)` to `torch.combinations(input, r=2, with_replacement=False, dim=0)`. 

For example, if the input tensor has shape `[n, d]` and `dim=0`, the resulting tensor would have shape `[n', d, r]`. The dimension with number of elements per combination (i.e. `r`) is added to the end. The value of  `n'` depends on `with_replacement` (e.g. would be n choose r in the case of `with_replacement=False`).

## Motivation
I have a tensor of n samples with d features each. I am interested in computing the Hodges-Lehmann estimate of the median on each feature (independently). I am not sure how to do this efficiently right now, but with the proposed addition, it would just be the following:

```python
matrix = ....  # n x d tensor
medians = torch.combinations(matrix).mean(-1).median(0).values
```

I'm sure there are many other use cases for this as well.

## Alternatives

If there is an efficient way to accomplish this with the current available methods, please let me know as it would make this feature request unnecessary.

The semantics on the default dimension and where the extra dimension is added could also changed based on convenience/efficiency concerns.

**Comment by zou3519** [comment]
Seems like a reasonable ask. We'd accept a PR implementing this.

**Comment by NickDienemann** [comment]
Hi, I have had a similar demand and put together the following solution. This is not a complete solution to the demanded feature but handles the following case correctly:
- combination_length (r in the questions text and in 1d pytorch implementation) is fixed to 2
- the dimension over which the combinations are calculated is always fixed to the last input dimension
- Besides the tensor whose combinations are supposed to be calculated, a mask has to be passed that is 1 at each position where an element is supposed to be used for combinations. To make this function compute combinations like the 1d-pytorch version the mask has to be set to 1 where a unique element is located (uniquenes wrt. to all other N elements in the same batch)

However, these limitations are merely for the sake of simplifying my implementation and not inherent limitations of the idea. I am confident that the implementation could be scaled to meet the complete features demands. 

Feel free to use this code to implement a more complete and more flexible solution.

```
def _batched_combinations(tensor,mask):
        """
        Calculates the combinations of the elements along the last dimension of input tensor. For elements that are masked (i.e. 0) in the mask, all output combination that would contain that element are set to 0.
        Also returns a mask indicating valid and invalid combinations(as defined by the input mask) in the output 

        Args:
            tensor (torch.tensor of size(*,N), where * is any number of trailing batch dimensions): tensors whose combinations along the last dim are supposed to be calulated
            mask (torch.tensor of size (*,N), where * is any number of trailing batch dimensions): tensor that contains 1 where a unique element is located
        Returns:
            (torch.tensor of size (*,max_num_combinations,2), where * is any number of trailing batch dimensions): contains all combinations of the input tensor
            (torch.tensor of size (*,max_num_combinations), where * is any number of trailing batch dimensions): boolean mask that contains 1s where a valid combination is stored in the first output
        """

        device,dtype = tensor.device,tensor.dtype
        batch_dimensions = list(tensor.size()[:-1])
        batch_dimensions_ones = [1]*len(batch_dimensions)
        N = tensor.size()[-1]

        #rewrite the input such that the first elements in each entity of the last dimension are the valid ones
        num_valid = mask.sum(dim=-1)            #size = (*batch_dimensions)
        is_valid_mask = torch.arange(0,N,device=device,dtype=dtype).reshape([1]*(mask.dim()-1)+ [-1]).repeat(*batch_dimensions,1) < num_valid.unsqueeze(-1).repeat(*batch_dimensions_ones,N)
        valid_indices = mask * torch.arange(0,N,device=device,dtype=torch.int64).reshape(*batch_dimensions_ones,-1).repeat(*batch_dimensions,1)    #size = (*batch_dimensions,N)
        valid_indices_ordered = torch.sort(valid_indices,dim=-1,descending=True).values
        valid_entries_at_beginning_tensor = torch.gather(tensor,dim=-1,index= valid_indices_ordered) *is_valid_mask #size = (*batch_dimensions,N)

        #calculate the combinations over all valid indices
        max_num_valid = int(num_valid.max())
        #handle special cases where no or only a single valid elements are passed
        if max_num_valid==0:
            return torch.zeros(batch_dimensions,device=device,dtype=dtype), torch.zeros(batch_dimensions,device=device,dtype=bool)
        elif max_num_valid ==1:
            return torch.zeros(batch_dimensions+[1,2],device=device,dtype=dtype), torch.zeros(batch_dimensions+[1],device=device,dtype=bool)
        num_combinations = int(math.factorial(max_num_valid)/(math.factorial(max_num_valid-2) *math.factorial(2)))
        valid_entries_at_beginning_tensor = valid_entries_at_beginning_tensor.unsqueeze(-2).repeat(*batch_dimensions_ones,num_combinations,1)
        index_combinations = torch.combinations(torch.arange(0,max_num_valid,device=device,dtype=torch.int64),r=2,with_replacement=False).reshape([1]*(valid_entries_at_beginning_tensor.dim()-2)+ [-1,2]).repeat(*batch_dimensions,1,1) #size(*batch_dimensions,num_valid.max(),2)
        combinations =torch.concat([
            torch.gather(valid_entries_at_beginning_tensor,dim=-1,index = index_combinations[...,(0,)]),
            torch.gather(valid_entries_at_beginning_tensor,dim=-1,index = index_combinations[...,(1,)])
        ],dim=-1)

        #calculate a mask that contains 1s for valid combinations and 0 for those combinations that serve as padding
        factorial = lambda x: torch.lgamma(x+1).exp() #pytorch implementation of factorial, see: https://discuss.pytorch.org/t/factorials-that-will-work-with-cuda/131716/2
        num_combinations_mask = factorial(num_valid)/(factorial(num_valid-2) *factorial(torch.full_like(num_valid,fill_value=2)))           #size= (*batch_dimensions)
        is_valid_combination_mask = torch.arange(0,num_combinations,dtype=torch.int64,device=device).reshape(*batch_dimensions_ones,-1).repeat(*batch_dimensions,1) < num_combinations_mask.unsqueeze(-1).repeat(*batch_dimensions_ones,num_combinations)                #
        
        #apply that mask to the combinations to filter invalid ones
        combinations = combinations * is_valid_combination_mask.unsqueeze(-1).repeat(*batch_dimensions_ones,1,2)
        return combinations, is_valid_combination_mask
``` 
        


Minimal working example:
```
tensor=torch.tensor([
                [1,2,3],
                [4,4,6]
          ])
mask = torch.tensor([
        [True,True,True],
        [True,False,True]
    ])

combinations, is_valid_combination_mask = _batched_combinations(tensor,mask)
```
Returns:
```
tensor([[[3, 2],
         [3, 1],
         [2, 1]],

        [[6, 4],
         [0, 0],
         [0, 0]]])

tensor([[ True,  True,  True],
        [ True, False, False]])
```

**Comment by mkrizmancic** [comment]
Hi, I had yet another similar requirement, and I stumbled upon this issue looking for existing solutions. In my case, I wanted to make combinations of a 1D tensor (using `r=2`, but any `r` can be supported), but instead of combinations between all elements, I needed "sparse" combinations between subsets of elements defined with a `batch` tensor.

For example, given the input tensor of `[1, 2, 3, 4, 5, 6, 7]`, batch tensor `[0, 0, 0, 1, 1, 1, 1]`, `r=2` and no replacements, I want combinations `[1, 2], [1, 3], [2, 3]` for elements belonging to group (batch) 0, and combinations `[4, 5], [4, 6], [4, 7], [5, 6], [5, 7], [6, 7]` for group (batch) 1.

This is inspired by how PyTorch Geometric handles batches, and learning on graphs was exactly what motivated this problem. This can be done easily with loops, but for large tensors and many iterations, the difference in execution time can be significant.

I know this is an old issue, but if somebody else runs into a similar problem, you can try using the solution below (the full annotated code is available [in this gist](https://gist.github.com/mkrizmancic/388aec7356c8a1de5166978fc70a48f3)):

```python
import torch
# to_dense_batch is a utility function from the PyTorch Geometric library (PyG).
# If you don't have it, you can install it with: pip install torch_geometric
from torch_geometric.utils import to_dense_batch

def batch_combinations(input_tensor, batch_index, r=2):
    """
    Generates combinations of a given size 'r' for elements in a tensor,
    constrained to batches specified by a batch index.

    Args:
        input_tensor (torch.Tensor): The input tensor of elements. Shape: [N, ...].
        batch_index (torch.Tensor): A tensor mapping each element in input_tensor
                                    to a batch. Shape: [N].
        r (int): The size of each combination.

    Returns:
        torch.Tensor: A tensor containing the combined values.
                      Shape: [r, num_total_combinations, ...].
    """
    device = input_tensor.device
    indices_global = torch.arange(input_tensor.shape[0], device=device)

    dense_indices, mask = to_dense_batch(indices_global, batch_index, fill_value=-1)

    num_options_per_batch = mask.sum(dim=1)
    max_options = num_options_per_batch.max().int()

    if max_options < r:
        return torch.empty((r, 0) + input_tensor.shape[1:],
                           dtype=input_tensor.dtype, device=device)

    template_combos = torch.combinations(torch.arange(max_options, device=device), r=r)
    global_candidates = dense_indices.T[template_combos].permute(2, 0, 1)

    largest_local_index = template_combos[:, -1] # Shape: [num_template_combos]
    combination_mask = largest_local_index < num_options_per_batch.unsqueeze(1)
    valid_combinations = global_candidates[combination_mask]
    combination_indices = valid_combinations.T

    return input_tensor[combination_indices]


if __name__ == '__main__':
    # Define sample input tensor and batch indices
    batch_index = torch.tensor([0, 0, 0, 1, 1, 1, 1, 2, 2])
    input_data = torch.tensor([10, 20, 30, 40, 50, 60, 70, 80, 90])

    # --- Test with r = 2 ---
    print("--- Testing with r = 2 ---")
    output_r2 = batch_combinations(input_data, batch_index, r=2)
    print("Output shape:", output_r2.shape)
    print("Output tensor:\n", output_r2)
    # Expected combinations:
    # Batch 0 (size 3): (10,20), (10,30), (20,30) -> 3 combos
    # Batch 1 (size 4): (40,50), (40,60), (40,70), (50,60), (50,70), (60,70) -> 6 combos
    # Batch 2 (size 2): (80,90) -> 1 combo
    # Total: 3 + 6 + 1 = 10 combinations. Shape should be [2, 10].

    print("\n" + "="*30 + "\n")

    # --- Test with r = 3 ---
    print("--- Testing with r = 3 ---")
    output_r3 = batch_combinations(input_data, batch_index, r=3)
    print("Output shape:", output_r3.shape)
    print("Output tensor:\n", output_r3)
    # Expected combinations:
    # Batch 0 (size 3): (10,20,30) -> 1 combo
    # Batch 1 (size 4): (40,50,60), (40,50,70), (40,60,70), (50,60,70) -> 4 combos
    # Batch 2 (size 2): No combos of size 3 possible.
    # Total: 1 + 4 + 0 = 5 combinations. Shape should be [3, 5].

    print("\n" + "="*30 + "\n")

    # --- Test with r = 4 ---
    print("--- Testing with r = 4 ---")
    output_r4 = batch_combinations(input_data, batch_index, r=4)
    print("Output shape:", output_r4.shape)
    print("Output tensor:\n", output_r4)
    # Expected combinations:
    # Batch 0 (size 3): No combos
    # Batch 1 (size 4): (40,50,60,70) -> 1 combo
    # Batch 2 (size 2): No combos
    # Total: 0 + 1 + 0 = 1 combination. Shape should be [4, 1].
```
